src/sysdolphin/baselib/fog.c

```diff
@@ -0,0 +1,256 @@
+
+#include <dolphin/gx/GX_unknown_001.h>
+#include <sysdolphin/baselib/cobj.h>
+
+static void FogInfoInit(void);
+static void FogAdjInfoInit(void);
+HSD_FogInfo hsdFog = { FogInfoInit };
+HSD_ClassInfo hsdFogAdj = { FogAdjInfoInit };
+
+typedef struct _GXFogAdjTbl {
+	u16 r[10];
+} GXFogAdjTbl;
+
+const GXColor lbl_804DE6F0 = { 0 };
+
+void HSD_FogSet(HSD_Fog* fog)
+{
+    GXFogAdjTbl sp88;
+    f32 sp70[6];
+    HSD_CObj* cobj;
+    s32 var_r31;
+    struct {
+        f32 x0;
+        f32 arr[6];
+    } sp54;
+
+    if (fog == NULL) {
+        GXSetFog(0, lbl_804DE6F0, 0.0F, 0.0F, 0.0F, 0.0F);
+        return;
+    }
+    cobj = HSD_CObjGetCurrent();
+    if (cobj == NULL) {
+        HSD_Panic(__FILE__, 0x4D, "You must specify CObj first.\n");
+    }
+    GXSetFog(fog->type, fog->color, fog->start, fog->end,
+        HSD_CObjGetNear(cobj), HSD_CObjGetFar(cobj));
+
+    if (fog->fog_adj != NULL) {
+        GXGetViewportv(sp70);
+        var_r31 = sp70[0] + sp70[2] * (fog->fog_adj->center + 320) / 640.0F;
+        if (var_r31 < 0) {
+            var_r31 = 0;
+        }
+        if (var_r31 > 640) {
+            var_r31 = 640;
+        }
+        if (fog->fog_adj->width != 0) {
+            GXInitFogAdjTable(&sp88, fog->fog_adj->width, fog->fog_adj->mtx);
+        } else {
+            Mtx44 sp14 = { 0 };
+            GXGetProjectionv(&sp54);
+            switch ((s32) sp54.x0) {
+            case 0:
+                sp14[0][0] = sp54.arr[0];
+                sp14[0][2] = sp54.arr[1];
+                sp14[1][1] = sp54.arr[2];
+                sp14[1][2] = sp54.arr[3];
+                sp14[2][2] = sp54.arr[4];
+                sp14[2][3] = sp54.arr[5];
+                sp14[3][2] = -1.0F;
+                break;
+            case 1:
+                sp14[0][0] = sp54.arr[0];
+                sp14[0][3] = sp54.arr[1];
+                sp14[1][1] = sp54.arr[2];
+                sp14[1][3] = sp54.arr[3];
+                sp14[2][2] = sp54.arr[4];
+                sp14[2][3] = sp54.arr[5];
+                sp14[3][3] = 1.0F;
+                break;
+            }
+            GXInitFogAdjTable(&sp88, sp70[2], sp14);
+        }
+        GXSetFogRangeAdj(1, var_r31, &sp88);
+        return;
+    }
+    GXSetFogRangeAdj(0, 0, NULL);
+}
+
+HSD_Fog* HSD_FogAlloc(void)
+{
+    HSD_Fog* fog = hsdNew((HSD_ClassInfo*) &hsdFog);
+    assert_line(0x8C, fog);
+    return fog;
+}
+
+HSD_Fog* HSD_FogLoadDesc(HSD_FogDesc* desc)
+{
+    HSD_Fog* fog = HSD_FogAlloc();
+    assert_line(0x99, fog);
+    HSD_FogInit(fog, desc);
+    if (desc->fogadjdesc != NULL) {
+        fog->fog_adj = HSD_FogAdjLoadDesc(desc->fogadjdesc);
+    }
+    return fog;
+}
+
+void HSD_FogInit(HSD_Fog* fog, HSD_FogDesc* desc)
+{
+    if (fog != NULL) {
+        if (desc != NULL) {
+            fog->type = desc->type;
+            fog->start = desc->start;
+            fog->end = desc->end;
+            fog->color = desc->color;
+        } else {
+            f32 sp10[6];
```

Comment:
Another autogenerated name. In this case it's `v` - you can extend that to viewport if you want to be explicit but I think the function name covers that.